internal/provider/resource_bgp_session.go

```diff
@@ -462,23 +465,43 @@ func extractBgpSessionCreate(d *schema.ResourceData) packetfabric.BgpSession {
 	return bgpSession
 }
 
-func extractBgpSessionUpdate(d *schema.ResourceData) packetfabric.BgpSession {
+func extractBgpSessionUpdate(d *schema.ResourceData, c *packetfabric.PFClient, cID string, connCID string) (packetfabric.BgpSession, error) {
 	bgpSession := packetfabric.BgpSession{}
 	if l3Address, ok := d.GetOk("l3_address"); ok {
 		bgpSession.L3Address = l3Address.(string)
 	}
-	// https://docs.packetfabric.com/api/v2/swagger/#/Cloud%20Router%20BGP%20Session%20Settings/cloud_routers_bgp_update
-	// Azure BGP session Update: l3_address = Azure Subnet (primary or secondary)
-	// set l3Address based on the values of primarySubnet and secondarySubnet when modified
-	// This is a temporary solution until the BGP API is refactored.
-	if d.HasChange("primary_subnet") {
-		if primarySubnet, ok := d.GetOk("primary_subnet"); ok {
-			bgpSession.L3Address = primarySubnet.(string)
-		}
+
+	if primarySubnet, ok := d.GetOk("primary_subnet"); ok {
+		bgpSession.PrimarySubnet = primarySubnet.(string)
 	}
-	if d.HasChange("secondary_subnet") {
-		if secondarySubnet, ok := d.GetOk("secondary_subnet"); ok {
-			bgpSession.L3Address = secondarySubnet.(string)
+	if secondarySubnet, ok := d.GetOk("secondary_subnet"); ok {
+		bgpSession.SecondarySubnet = secondarySubnet.(string)
+	}
+
+	// For Azure, the L3Address is not set so determine which subnet to use
+	if "" == bgpSession.L3Address {
+		crConnection, err := c.ReadCloudRouterConnection(cID, connCID)
+		if err != nil {
+			return packetfabric.BgpSession{}, err
+		}
+		connectionType := crConnection.CloudSettings.AzureConnectionType
```

Comment:
We should probably first check the connection type and make sure it is indeed an azure connection and if not, return an error saying l3_address is required.

Reply:
Good point! I'll add it